### PR TITLE
feat(vse): added a vse domain factory

### PR DESCRIPTION
**Pull request by easen-amp**
that can generate a vse domain using the supplied snapshotId or timestamp

### CODE REVIEW

**Comment by easen-amp** [comment]
> Great addition. Could be added to readme to make it more discoverable.

Thanks @pixelscript. I've added a new section and included some examples :)

After speaking to @techiedarren I've renamed this new factory to be StagingEnvironmentFactory, as it in keeping with the existing configuration property that you pass into `ContentClient` (`ContentClient({stagingEnvironment: ...})`)

File: README.md [hunk 1]
```diff
@@ -126,17 +126,43 @@ Example:
...
 ### Preview staging content
 
-By default, the content client will request content from the production content delivery services. When a user wants to preview content before it is published you can repoint the client to a virtual staging environment (VSE):
+By default, the content client will request content from the production content delivery services. When a user wants to preview content before it is published you can re-point the client to a virtual staging environment (VSE):
 
 ```js
 var client = new ContentClient({
     account: 'myaccount',
-    stagingEnvironment: 'fhboh562c3tx1844c2ycknz96-gvzrfgnzc-1546264721816.staging.bigcontent.io'
+    stagingEnvironment: 'fhboh562c3tx1844c2ycknz96.staging.bigcontent.io'
```
Comment (easen-amp) [inline README.md]: removed the timestamp componented from the domain

File: src/lib/staging-environment/StagingEnvironmentFactory.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+   * Generates a new staging environment domain name using the snapshotId and/or timestamp thats supplied in the options argument
+   * @param options
+   */
+  generateDomain(options: GenerateDomainOptions): Promise<string> {
+    const url = this.buildUrl(options);
+
+    return this.client
+      .get<string>(url)
+      .then(response => {
+        return Promise.resolve(response.data);
```
Comment (dev-warner) [inline src/lib/staging-environment/StagingEnvironmentFactory.ts]: You don't need to return a promise in a then it already returns a promise. also best to use async await :)

File: src/lib/staging-environment/StagingEnvironmentFactory.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      })
+      .catch(err => {
+        return Promise.reject(
```
Comment (dev-warner) [inline src/lib/staging-environment/StagingEnvironmentFactory.ts]: use `throw new Error` instead

File: README.md [hunk 4]
```diff
@@ -126,17 +126,43 @@ Example:
...
+
+var client = new ContentClient({
+  account: 'myaccount',
+  stagingEnvironment: stagingEnvironmentAtSnapshot
+});
+```
+
+Previewing content at a given timestamp (epoch milliseconds):
+```js
+var factory = new StagingEnvironmentFactory('fhboh562c3tx1844c2ycknz96.staging.bigcontent.io');
```
Comment (dev-warner) [inline README.md]: use `const` instead of var in examples

Comment (easen-amp) [reply README.md]: This was because all of the other examples were using`var`, I assume this is because we may need to support older browsers.

Comment (dev-warner) [reply README.md]: all browsers support const https://caniuse.com/#search=const

Comment (dev-warner) [reply README.md]: it's not a big issue just rarely ever see var used in examples and makes it appear outdated imo.  JS classes have less support than const so we should be encouraging build steps over using old standards for browser support.

Comment (easen-amp) [reply README.md]: I'll make another PR for the var --> const change, just to keep it seperate.

**Pull request merged**